src/rust/lib/flexer/src/automata/nfa.rs

```diff
@@ -0,0 +1,163 @@
+        self.states[source].links.add(target)
+    }
+
+
+    //// NFA -> DFA ////
+
+    fn fill_eps_matrix(&self, state_to_mat:&mut Vec<EpsMatrix>, i:usize) {
+        fn go(
+            nfa              : &NFA,
+            state_to_mat     : &mut Vec<EpsMatrix>,
+            unitialized      : &mut Vec<bool>,
+            eps_group_ix_map : &mut HashMap<HashSet<usize>, usize>,
+            i                : usize,
+        ){
+            let mut eps_links = HashSet::<usize>::with_capacity(i);
+            if unitialized[i] {
+                let mut circular   = false;
+                let mut eps_matrix = EpsMatrix::default();
+                state_to_mat[i] = eps_matrix;
+                unitialized[i]  = false;
+                for tgt in nfa.states[i].links.epsilon {
+                    go(nfa,state_to_mat,unitialized,eps_group_ix_map,tgt);
+                    eps_links.insert(tgt);
+                    eps_links.extend(state_to_mat[tgt].links);
+                    if !state_to_mat[tgt].computed {
+                        circular = true
+                    }
+                }
+                if !circular {
+                    if eps_group_ix_map.get(&eps_links).is_none() {
+                        eps_group_ix_map[eps_links] = eps_group_ix_map.len();
+                    }
+                    eps_matrix.computed = true
+                }
+                eps_matrix.links = eps_links.clone();
+            }
+        }
+        let mut eps_group_ix_map = HashMap::new();
+        let mut uninitialized    = vec![true;state_to_mat.len()];
+        go(self,state_to_mat,&mut uninitialized,&mut eps_group_ix_map,i);
+    }
+
+    fn eps_matrix(&self) -> Vec<HashSet<i64>> {
+        let mut arr = vec![EpsMatrix::default(); self.states.len()];
+        for state_ix in 0..self.states.len() {
+            self.fill_eps_matrix(&mut arr, state_ix);
+        }
+        arr.into_iter().map(|m| m.links).collect()
+    }
+
+    fn nfa_matrix(&self) -> Vec<Vec<i64>> {
+        //    logger.group("Computing NFA Matrix")
+        let matrix = vec![vec![0; self.states.len()]; self.vocabulary.len()];
+        for stateIx in self.states.indices {
+            let source = self.states[stateIx];
+            for (range, vocIx) in self.vocabulary {
+                let target = match source.links.ranged.get(range.start) {
+                    Some(target) => target,
+                    None         => state::MISSING,
+                };
+                matrix[stateIx][vocIx] = target;
+            }
+        }
+        matrix
+    }
+
+    pub fn into_dfa(self) -> DFA {
+        let     vocabulary  = self.vocabulary;
+        let     eps_mat     = self.eps_matrix();
+        let     nfa_mat     = self.nfa_matrix();
+        let mut dfa_rows    = 0;
+        let mut dfa_mat     = Vec::<Vec<i64>>::new();
+        let mut dfa_eps_ixs = Vec::<HashSet<i64>>::new();
+        let mut dfa_eps_map = HashMap::<HashSet<i64>,i64>::new();
```

Comment:
HashSet does not implement Hash in rust, so we cannot use it as key. Could we use the hashset index in `dfa_eps_ixs` instead?